src/material/schematics/ng-generate/mdc-migration/index.ts

```diff
@@ -55,48 +61,90 @@ function getComponentsToMigrate(requested: string[]): Set<string> {
+  fileSystem: DevkitFileSystem,
+  tsconfigPath: WorkspacePath,
+  migrators: ComponentMigrator[],
+  analyzedFiles: Set<WorkspacePath>,
+  additionalStylesheetPaths: string[],
+): boolean {
+  const program = UpdateProject.createProgramFromTsconfig(tsconfigPath, fileSystem);
+  const project = new UpdateProject(context, program, fileSystem, analyzedFiles, context.logger);
+  return !project.migrate(
+    [ThemingStylesMigration, TemplateMigration, RuntimeCodeMigration],
+    null,
+    migrators,
+    additionalStylesheetPaths,
+  ).hasFailures;
+}
 
-  const migrators: ComponentMigrator[] = [];
-  for (let i = 0; i < MIGRATORS.length; i++) {
-    if (componentsToMigrate.has(MIGRATORS[i].component)) {
-      migrators.push(MIGRATORS[i]);
+export default function (options: Schema): Rule {
+  return async (tree: Tree, context: SchematicContext) => {
+    const logger = context.logger;
+    const workspace = await getWorkspaceConfigGracefully(tree);
+    if (workspace === null) {
+      logger.error('Could not find workspace configuration file.');
+      return;
     }
-  }
 
-  return (tree: Tree, context: SchematicContext) => {
+    const projectNames = workspace.projects.keys();
     const fileSystem = new DevkitFileSystem(tree);
-    const program = UpdateProject.createProgramFromTsconfig(
-      fileSystem.resolve(tsconfigPath),
-      fileSystem,
-    );
+    const analyzedFiles = new Set<WorkspacePath>();
+    const componentsToMigrate = getComponentsToMigrate(options.components);
+    const migrators = MIGRATORS.filter(m => componentsToMigrate.has(m.component));
+    let additionalStylesheetPaths = options.directory
+      ? findStylesheetFiles(tree, options.directory)
+      : [];
+    let success = true;
 
-    const additionalStylesheetPaths = findStylesheetFiles(tree, migrationDir);
-    const project = new UpdateProject(context, program, fileSystem, new Set(), context.logger);
-    const {hasFailures} = project.migrate(
-      [ThemingStylesMigration, TemplateMigration, RuntimeCodeMigration],
-      null,
-      migrators,
-      additionalStylesheetPaths,
-    );
+    logger.info(`Migrating components:\n${[...componentsToMigrate].join('\n')}`);
+
+    for (const projectName of projectNames) {
+      const project = workspace.projects.get(projectName)!;
+      const tsconfigPaths = [
+        getTargetTsconfigPath(project, 'build'),
+        getTargetTsconfigPath(project, 'test'),
+      ].filter((p): p is WorkspacePath => !!p);
+
+      if (!tsconfigPaths.length) {
+        logger.warn(
+          `Skipping migration for project ${projectName}. Unable to determine 'tsconfig.json' file in workspace config.`,
+        );
+        continue;
+      }
+
+      if (!options.directory) {
+        additionalStylesheetPaths = findStylesheetFiles(tree, project.root);
+      }
+
+      logger.info(`Migrating project: ${projectName}`);
+
+      success =
+        success &&
+        tsconfigPaths
+          .map(tsconfigPath =>
```

Comment:
nit: the side-effect-ey `map` with an `every` here seems a little weird to me since I'd expect a `map` to be pure.

Reply:
yeah fair, I'll change that to a for...of